tool/tsh/recording_export.go

```diff
@@ -0,0 +1,220 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"image"
+	"image/draw"
+	"image/jpeg"
+	"image/png"
+	"os"
+
+	"github.com/gravitational/trace"
+	"github.com/icza/mjpeg"
+
+	apievents "github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/session"
+	"github.com/gravitational/teleport/lib/srv/desktop/tdp"
+)
+
+const (
+	framesPerSecond  = 30
+	frameDelayMillis = float64(1000) / framesPerSecond
+)
+
+func onExportRecording(cf *CLIConf) error {
+	tc, err := makeClient(cf, false)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	proxyClient, err := tc.ConnectToProxy(cf.Context)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer proxyClient.Close()
+
+	authClient := proxyClient.CurrentCluster()
+	defer authClient.Close()
+
+	fname := cf.OutFile
+	if fname == "" {
+		fname = cf.SessionID + ".avi"
+	}
+
+	frames, err := writeMovie(cf.Context, authClient, session.ID(cf.SessionID), fname)
+	// there may be a partial file, even if we encountered an error,
+	// so we indicate to the user when we wrote something
+	if frames > 0 {
+		fmt.Printf("wrote recoring to %v\n", fname)
+	}
+
+	return trace.Wrap(err)
+}
+
+// writeMovie writes streams the events for the specified session into a movie file
+// identified by fname. It returns the number of frames that were written and an error.
+func writeMovie(ctx context.Context, ss events.SessionStreamer, sid session.ID, fname string) (int, error) {
+	var screen *image.NRGBA
+	var movie mjpeg.AviWriter
+
+	lastEmitted := int64(-1)
+	buf := new(bytes.Buffer)
+	frameCount := 0
+
+	evts, errs := ss.StreamSessionEvents(ctx, sid, 0)
+loop:
+	for {
+		select {
+		case err := <-errs:
+			if movie != nil {
+				movie.Close()
+				os.Remove(fname)
```

Comment:
Should we be removing the file (and on context below)? Comment above says there will be partial file even if we encounter an error (and I guess it's also useful to have partial file if user ctrl+c it)

Reply:
Yep, great catch. This is a leftover from before I decided that a partial file makes more sense.